tests/test_operation.py

```diff
@@ -177,6 +177,44 @@ def test_operation_init(self, test_class, monkeypatch):
         if n == 0:
             return
 
+    def test_barrier(self):
+        r"Test that the barrier influences compilation"
+
+        def qfunc():
+            qml.Hadamard(wires=0)
+            qml.Barrier(wires=0)
+            qml.Hadamard(wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        dev = qml.device("default.qubit", wires=3)
+        qnode = qml.QNode(qfunc, dev)
+        gates = qml.specs(qnode)()["gate_sizes"][1]
+
+        assert gates == 3
+
+        optimized_qfunc = qml.compile()(qfunc)
+        optimized_qnode = qml.QNode(optimized_qfunc, dev)
+        optimized_gates = qml.specs(optimized_qnode)()["gate_sizes"][1]
+
+        assert optimized_gates == 2
+
+        # check only_visual parameter
```

Comment:
This could be a separate test case.
```suggestion
    def test_barrier_only_visual(self):
        r"""Test that the barrier doesn't influence compilation when the only_visual parameter is True."""
```